--- a/gprofng/libcollector/iotrace.c
+++ b/gprofng/libcollector/iotrace.c
@@ -1030,8 +1030,9 @@
   }
 
 DCL_FUNC_VER (DCL_OPEN64, open64_2_2, open64@GLIBC_2.2)
+#if !defined(_LARGEFILE64_SOURCE) && !defined(_GNU_SOURCE)
 DCL_OPEN64 (open64)
+#endif
-
 
 #define F_ERROR_ARG     0
 #define F_INT_ARG       1
@@ -2989,7 +2990,9 @@
 DCL_FUNC_VER (DCL_FGETPOS64, fgetpos64_2_2_5, fgetpos64@GLIBC_2.2.5)
 DCL_FUNC_VER (DCL_FGETPOS64, fgetpos64_2_2, fgetpos64@GLIBC_2.2)
 DCL_FUNC_VER (DCL_FGETPOS64, fgetpos64_2_1, fgetpos64@GLIBC_2.1)
+#if !defined(_LARGEFILE64_SOURCE) && !defined(_GNU_SOURCE)
 DCL_FGETPOS64 (fgetpos64)
+#endif
 
 /*------------------------------------------------------------- fsetpos */
 static int
@@ -3087,7 +3089,9 @@
 DCL_FUNC_VER (DCL_FSETPOS64, fsetpos64_2_2_5, fsetpos64@GLIBC_2.2.5)
 DCL_FUNC_VER (DCL_FSETPOS64, fsetpos64_2_2, fsetpos64@GLIBC_2.2)
 DCL_FUNC_VER (DCL_FSETPOS64, fsetpos64_2_1, fsetpos64@GLIBC_2.1)
+#if !defined(_LARGEFILE64_SOURCE) && !defined(_GNU_SOURCE)
 DCL_FSETPOS64 (fsetpos64)
+#endif
 
 /*------------------------------------------------------------- fsync */
 int
