--- a/include/dlfcn.h
+++ b/include/dlfcn.h
@@ -36,7 +36,12 @@
 #endif
 
 #if _REDIR_TIME64
+#ifndef NO_GLIBC_ABI_COMPATIBLE
+__attribute__((__weak__)) void *__dlsym_time64(void *__restrict handle, const char *__restrict symbol) { return dlsym(handle, symbol); }
+#define dlsym __dlsym_time64
+#else
 __REDIR(dlsym, __dlsym_time64);
+#endif
 #endif
 
 #ifdef __cplusplus
--- a/include/pthread.h
+++ b/include/pthread.h
@@ -198,6 +198,13 @@
 
 int pthread_atfork(void (*)(void), void (*)(void), void (*)(void));
 
+#ifndef NO_GLIBC_ABI_COMPATIBLE
+extern void *__dso_handle __attribute__((__weak__, __visibility__("hidden")));
+
+int __register_atfork(void (*)(void), void (*)(void), void (*)(void), void *);
+#define pthread_atfork(prepare, parent, child) __register_atfork(prepare, parent, child, &__dso_handle ? __dso_handle : 0)
+#endif
+
 int pthread_getconcurrency(void);
 int pthread_setconcurrency(int);
 
@@ -207,6 +214,7 @@
 	void (*__f)(void *);
 	void *__x;
 	struct __ptcb *__next;
+	void *__unused;
 };
 
 void _pthread_cleanup_push(struct __ptcb *, void (*)(void *), void *);
--- a/include/setjmp.h
+++ b/include/setjmp.h
@@ -27,6 +27,9 @@
 typedef jmp_buf sigjmp_buf;
 int sigsetjmp (sigjmp_buf, int) __setjmp_attr;
 _Noreturn void siglongjmp (sigjmp_buf, int);
+#ifndef NO_GLIBC_ABI_COMPATIBLE
+__REDIR(sigsetjmp, __sigsetjmp);
+#endif
 #endif
 
 #if defined(_XOPEN_SOURCE) || defined(_GNU_SOURCE) \
--- a/include/stdlib.h
+++ b/include/stdlib.h
@@ -50,6 +50,18 @@
 int at_quick_exit (void (*) (void));
 _Noreturn void quick_exit (int);
 
+#ifndef NO_GLIBC_ABI_COMPATIBLE
+extern void *__dso_handle __attribute__((__weak__, __visibility__("hidden")));
+typedef void (*__cxa_func)(void *);
+static inline __cxa_func __cxa_func_convert(void (*func)(void)) { return (__cxa_func)func; }
+
+int __cxa_atexit(__cxa_func, void *, void *);
+__attribute__((__weak__)) int atexit(void (*func)(void)) { return __cxa_atexit(__cxa_func_convert(func), 0, &__dso_handle ? __dso_handle : 0); }
+
+int __cxa_at_quick_exit(__cxa_func, void *);
+#define at_quick_exit(func) __cxa_at_quick_exit(__cxa_func_convert(func), &__dso_handle ? __dso_handle : 0)
+#endif
+
 char *getenv (const char *);
 
 int system (const char *);
@@ -161,6 +173,11 @@
 float strtof_l(const char *__restrict, char **__restrict, struct __locale_struct *);
 double strtod_l(const char *__restrict, char **__restrict, struct __locale_struct *);
 long double strtold_l(const char *__restrict, char **__restrict, struct __locale_struct *);
+
+#ifndef NO_GLIBC_ABI_COMPATIBLE
+__attribute__((__weak__)) char *__secure_getenv(const char *name) { return getenv(name); }
+__attribute__((__weak__)) char *secure_getenv(const char *name) { return __secure_getenv(name); }
+#endif
 #endif
 
 #if defined(_LARGEFILE64_SOURCE)
--- a/src/env/secure_getenv.c
+++ b/src/env/secure_getenv.c
@@ -6,3 +6,5 @@
 {
 	return libc.secure ? NULL : getenv(name);
 }
+
+weak_alias(secure_getenv, __secure_getenv);
--- a/src/errno/__errno_location.c
+++ b/src/errno/__errno_location.c
@@ -1,9 +1,9 @@
 #include <errno.h>
 #include "pthread_impl.h"
 
-int *__errno_location(void)
+__attribute__((const)) hidden int *___errno_location(void)
 {
 	return &__pthread_self()->errno_val;
 }
 
-weak_alias(__errno_location, ___errno_location);
+weak_alias(___errno_location, __errno_location);
--- a/src/exit/at_quick_exit.c
+++ b/src/exit/at_quick_exit.c
@@ -22,12 +22,19 @@
 	}
 }
 
-int at_quick_exit(void (*func)(void))
+int __cxa_at_quick_exit(void (*func)(void *), void *dso)
 {
 	int r = 0;
 	LOCK(lock);
 	if (count == 32) r = -1;
-	else funcs[count++] = func;
+	else funcs[count++] = (void (*)(void))func;
 	UNLOCK(lock);
 	return r;
 }
+
+extern void *__dso_handle __attribute__((__weak__, __visibility__("hidden")));
+
+int at_quick_exit(void (*func)(void))
+{
+	return __cxa_at_quick_exit((void (*)(void *))(func), &__dso_handle ? __dso_handle : 0);
+}
--- a/src/include/errno.h
+++ b/src/include/errno.h
@@ -3,12 +3,4 @@
 
 #include "../../include/errno.h"
 
-#ifdef __GNUC__
-__attribute__((const))
-#endif
-hidden int *___errno_location(void);
-
-#undef errno
-#define errno (*___errno_location())
-
 #endif
--- a/src/thread/__lock.c
+++ b/src/thread/__lock.c
@@ -19,7 +19,7 @@
 void __lock(volatile int *l)
 {
 	int need_locks = libc.need_locks;
-	if (!need_locks) return;
+	if (!need_locks && libc.auxv) return;
 	/* fast path: INT_MIN for the lock, +1 for the congestion */
 	int current = a_cas(l, 0, INT_MIN + 1);
 	if (need_locks < 0) libc.need_locks = 0;
--- a/src/thread/pthread_atfork.c
+++ b/src/thread/pthread_atfork.c
@@ -53,3 +53,8 @@
 	UNLOCK(lock);
 	return 0;
 }
+
+int __register_atfork(void (*prepare)(void), void (*parent)(void), void (*child)(void), void *dso_handle)
+{
+	return pthread_atfork(prepare, parent, child);
+}
--- a/src/time/clock_gettime.c
+++ b/src/time/clock_gettime.c
@@ -3,6 +3,7 @@
 #include <stdint.h>
 #include "syscall.h"
 #include "atomic.h"
+#include "libc.h"
 
 #ifdef VDSO_CGT_SYM
 
@@ -33,7 +34,9 @@
 
 static int cgt_init(clockid_t clk, struct timespec *ts)
 {
-	void *p = __vdsosym(VDSO_CGT_VER, VDSO_CGT_SYM);
+	void *p = (void *)0;
+	if (libc.auxv) {
+	p = __vdsosym(VDSO_CGT_VER, VDSO_CGT_SYM);
 #ifdef VDSO_CGT32_SYM
 	if (!p) {
 		void *q = __vdsosym(VDSO_CGT32_VER, VDSO_CGT32_SYM);
@@ -46,6 +49,7 @@
 	if (!__vdsosym(VDSO_CGT32_VER, VDSO_CGT32_SYM)) p = 0;
 #endif
 #endif
+	}
 	int (*f)(clockid_t, struct timespec *) =
 		(int (*)(clockid_t, struct timespec *))p;
 	a_cas_p(&vdso_func, (void *)cgt_init, p);
