--- a/libgcc/Makefile.in
+++ b/libgcc/Makefile.in
@@ -247,7 +247,8 @@
 LIBGCC2_CFLAGS = -O2 $(LIBGCC2_INCLUDES) $(GCC_CFLAGS) $(HOST_LIBGCC2_CFLAGS) \
 		 $(LIBGCC2_DEBUG_CFLAGS) -DIN_LIBGCC2 \
 		 -fbuilding-libgcc -fno-stack-protector \
-		 $(INHIBIT_LIBC_CFLAGS)
+		 $(INHIBIT_LIBC_CFLAGS) \
+		 -g0 -Os -DNO_GLIBC_ABI_COMPATIBLE
 
 # Additional options to use when compiling libgcc2.a.
 # Some targets override this to -isystem include
