--- a/libstdc++-v3/configure.ac
+++ b/libstdc++-v3/configure.ac
@@ -123,6 +123,11 @@
   glibcxx_compiler_pic_flag="$lt_prog_compiler_pic_CXX"
   glibcxx_compiler_shared_flag="-D_GLIBCXX_SHARED"
 
+elif test "${with_pic+set}" = set; then
+  glibcxx_lt_pic_flag="-prefer-pic"
+  glibcxx_compiler_pic_flag="$lt_prog_compiler_pic_CXX"
+  glibcxx_compiler_shared_flag=
+
 else
   glibcxx_lt_pic_flag=
   glibcxx_compiler_pic_flag=
--- a/libstdc++-v3/configure
+++ b/libstdc++-v3/configure
@@ -15603,8 +15603,8 @@
   glibcxx_compiler_shared_flag="-D_GLIBCXX_SHARED"
 
 else
-  glibcxx_lt_pic_flag=
-  glibcxx_compiler_pic_flag=
+  glibcxx_lt_pic_flag="-prefer-pic"
+  glibcxx_compiler_pic_flag="$lt_prog_compiler_pic_CXX"
   glibcxx_compiler_shared_flag=
 fi
 
